src/Oscoin/Consensus/Mining.hs

```diff
@@ -34,21 +35,23 @@ mineBlock
        , AuthTree.MerkleHash (Crypto.Hash c)
        )
     => BlockStore c tx s m
+    -> (Block c tx (Sealed c s) -> m ())
```

Comment:
You are right. When I was writing some glue code for the tests, at some point I did ask to myself: wait, `mineBlock` is _already_ returning the mined block, so cannot I simply use it and do whatever I want with it without even passing the continuation here?

However, after I did refactor the code to get rid of it, I stopped and unwinded my work, mostly because it occurred to me that now is responsibility of _the caller_ to be sure the block is persisted, vs responsibility of just the `mineBlock` function. I am thinking here that perhaps it's a good thing if, as part of the `mineBlock` contract, we can guarantee that we write such logic in just one place, as it will shelter us from programmer's errors (i.e. forget to do the "right thing" with the block). What do you think?

Reply:
I think that we should appreciate the fact that the `Node` is our "App Monad", which is basically omniscient and wires up all the different components (perhaps we could even start to consider moving more of `main` into `Node.run`).

Since the `Node` controls the mining loop, and also knows about storage resp. protocol, we'd end up with exactly one place to wire the two together.

Reply:
That's a good observation. The _only_ thing we should check is whether or not the execution order is important. Currently we have (inside `Consensus.Mining`):

```haskell
            for maybeSealedBlock $ \blk -> do
                Mempool.delTxs (blockData blk)
                dispatchBlock blk
                StateStore.storeState st'
                for_ receipts addReceipt
                pure blk
```

If `storeState` and `addReceipt` have as precondition that the block must be stored then we might need to keep the (synchronous) callback, otherwise doing this in `Node.mineBlock` is perfectly OK. I will try to assume the order is independent, and hopefully tests should catch any regression I might have introduced.

Reply:
Ah fair point!